File: sol/solution/0001-0100/0020/Solution.kt
class Solution {
    fun isValid(s: String): Boolean {
        val stack = mutableListOf<Char>()
        for (c in s) {
            if (c == '(' || c == '{' || c == '[') {
                stack.add(c)
            } else {
                if (stack.isEmpty()) return false
                val top = stack.removeAt(stack.size - 1)
                if ((c == ')' && top != '(') || 
                    (c == '}' && top != '{') || 
                    (c == ']' && top != '[')) {
                    return false
                }
            }
        }
        return stack.isEmpty()
    }
}

File: sol/solution/0001-0100/0020/Solution.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = Vec::new();
        for c in s.chars() {
            if c == '(' || c == '{' || c == '[' {
                stack.push(c);
            } else {
                if stack.is_empty() {
                    return false;
                }
                let top = stack.pop().unwrap();
                if (c == ')' && top != '(') || 
                   (c == '}' && top != '{') || 
                   (c == ']' && top != '[') {
                    return false;
                }
            }
        }
        stack.is_empty()
    }
}